editor/gui/scene_tree_editor.cpp

```diff
@@ -1486,8 +1487,48 @@ void SceneTreeDialog::popup_scenetree_dialog() {
 	popup_centered_clamped(Size2(350, 700) * EDSCALE);
 }
 
+void SceneTreeDialog::set_valid_types(const Vector<StringName> &p_valid) {
+	if (p_valid.is_empty()) {
+		return;
+	}
+
+	tree->set_valid_types(p_valid);
+
+	HBoxContainer *hbox = memnew(HBoxContainer);
+	content->add_child(hbox);
+	content->move_child(hbox, 0);
+
+	{
+		Label *label = memnew(Label);
+		hbox->add_child(label);
+		label->set_text(TTR("Allowed:"));
+	}
+
+	HFlowContainer *hflow = memnew(HFlowContainer);
+	hbox->add_child(hflow);
+	hflow->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+
+	for (const StringName &type : p_valid) {
+		TextureRect *trect = memnew(TextureRect);
+		hflow->add_child(trect);
+		trect->set_expand_mode(TextureRect::EXPAND_IGNORE_SIZE);
+		trect->set_stretch_mode(TextureRect::STRETCH_KEEP_ASPECT_CENTERED);
+		trect->set_meta("type", type);
+		valid_type_icons.push_back(trect);
+
+		Label *label = memnew(Label);
+		hflow->add_child(label);
+		label->set_text(type);
+		label->set_auto_translate(false);
```

Comment:
We should group them in hbox, otherwise the label and the icon can end up on separate lines.

![image](https://github.com/godotengine/godot/assets/11782833/95f1749c-e79a-495c-9670-8c143b854370)